eth/get_proof.go

```diff
@@ -33,17 +36,41 @@ func (n *proofList) Delete(key []byte) error {
 	panic("not supported")
 }
 
-func NewProofProvider() *ProofProvider {
-	return &ProofProvider{}
+func NewProofProvider(db state.Database, blockStore store.BlockStoreReader) *ProofProvider {
+	return &ProofProvider{
+		database:   db,
+		blockStore: blockStore,
+	}
 }
 
 // getState returns the state.StateBD and Header of block at the given number.
 // If the passed number is nil, it returns the the db and header of the latest block.
 //
-// TODO: replace this with a real implementation based on `ProofProvider` being supplied
+// WIP: replace this with a real implementation based on `ProofProvider` being supplied
 // with required data sources
 func (p *ProofProvider) getState(blockNumber *big.Int) (*state.StateDB, types.Header, error) {
-	return nil, types.Header{}, ErrNotImplemented
+	var ethBlock *types.Block
+	var err error
+
+	if blockNumber == nil {
+		ethBlock, err = p.blockStore.HeadBlock().ToEth()
+	} else {
+		ethBlock, err = p.blockStore.BlockByNumber(blockNumber.Int64()).ToEth()
+	}
+
+	if err != nil {
+		return nil, types.Header{}, fmt.Errorf("getting eth block %d: %w", blockNumber, err)
+	}
+
+	header := ethBlock.Header()
+	hash := ethBlock.Hash()
+
+	sdb, err := state.New(hash, p.database, nil)
+	if err != nil {
+		return nil, *header, fmt.Errorf("opening state.StateDB: %w", err)
+	}
```

Comment:
Since the ethereum `StateRoot` is stored on the monomer block header, the block shouldn't need to be converted to an eth block. I think we can just pull `StateRoot` directly off of the monomer block header and use it to build the eth state 

```suggestion
	var block *types.Block
	var err error

	if blockNumber == nil {
		block, err = p.blockStore.HeadBlock()
	} else {
		block, err = p.blockStore.BlockByNumber(blockNumber.Int64())
	}

	if err != nil {
		return nil, types.Header{}, fmt.Errorf("getting block %d: %w", blockNumber, err)
	}

	header := block.Header()

	sdb, err := state.New(header.StateRoot, p.database, nil)
	if err != nil {
		return nil, *header, fmt.Errorf("opening state.StateDB: %w", err)
	}
```

Reply:
The `header` produced in this change is a `monomer.Header` but the (copied) code in `GetProof` is wanting a `geth.core.types.Header`.

Consuming code only really wants the `Root` from this header, so I could change to return that directly from the monomer header, but I'm more inclined to minimize the diff in the copied functions, at least until some tests are in place.

Reply:
Ah ok I see your point, we can always revisit this during testing. In that case could you use `ethBlock.Root()` instead of `ethBlock.Hash()`? I think that has the eth state root that we want to use in the state db.

Reply:
> I'm more inclined to minimize the diff in the copied functions, at least until some tests are in place.

Agreed here. Helps with future-proofing too :+1:

Reply:
Sounds good, I'll go ahead and approve for now and then we can revisit this tomorrow once we have the withdrawals in eth storage